src/NUnitFramework/framework/Internal/TestExecutionContext.cs

```diff
@@ -493,6 +494,28 @@ public void AddFormatter(ValueFormatterFactory formatterFactory)
             CurrentValueFormatter = formatterFactory(CurrentValueFormatter);
         }
 
+        /// <summary>
+        /// Attach a file to the current test result
+        /// </summary>
+        /// <param name="filepath">Filepath to attachment. Can be absolute, or relative to working directory.</param>
+        /// <param name="description">Optional description of attachment</param>
+        public void AddTestAttachment(string filepath, string description = null)
+        {
+            Guard.ArgumentNotNull(filepath, nameof(filepath));
+            Guard.ArgumentValid(filepath.IndexOfAny(Path.GetInvalidPathChars()) == -1, 
+                $"Test attachment filepath contains invalid path characters. {filepath}", nameof(filepath));
+
+            if (!Path.IsPathRooted(filepath))
+                filepath = Path.Combine(TestContext.CurrentContext.WorkDirectory, filepath);
+
+#if !PORTABLE
+            if (!File.Exists(filepath))
+                throw new FileNotFoundException("Test attachment filepath could not be found.", filepath);
```

Comment:
This was a point I went back and forth on. 🙂 My thinking was that adding a non-existant file is far more likely to occur through user-error, rather than intention, and you'd want to know about that ASAP. I imagine a common use case may be where people only refer to attachments when the test has failed - so broken paths may only be discovered 6 months after writing, when the test first fails, and it's too late.

The opposite arguments e.g. 'adding an attachment before saving it', mostly seem like things that could be managed by the user. (e.g. Don't add an attachment before you've created it!) That seems like fairly bad practice anyway, what if the creation of the attachment later went wrong, and the file couldn't be written? We could always add a 'don't check file existence' overload in the future if such cases arose.

Not that it matters, but MSTest checks if the File Exists. @rprouse @jnm2 - what do you guys think?